File: analysis/ghidra_scripts/SeedUpdatePath.py
## ###
#  IP: GHIDRA
##
# Seed likely update-path routines in a raw JJOS firmware image.
# @category: Analysis
# @runtime Jython

from ghidra.app.cmd.disassemble import DisassembleCommand
from ghidra.app.util import PseudoDisassembler
from jarray import zeros

BASE = 0x09000000

RANGES = [
    ("boot_update_block", BASE + 0x7be0, BASE + 0x8130, False),
    ("progress_block", BASE + 0x94a0, BASE + 0x94d0, False),
    ("progress_erase_exact", BASE + 0x94a0, BASE + 0x94a1, True),
    ("progress_write_exact", BASE + 0x94b8, BASE + 0x94b9, True),
]


def safe_create_label(addr, name):
    try:
        if getSymbolAt(addr) is None:
            createLabel(addr, name, True)
    except:
        pass


def create_anchor_labels():
    safe_create_label(toAddr(BASE + 0x7c9c), "STR_BOOT_OS_UPDATE_TITLE")
    safe_create_label(toAddr(BASE + 0x7cf8), "STR_BOOT_LOADING_OS_FILE")
    safe_create_label(toAddr(BASE + 0x7d14), "STR_BOOT_OS_FILE_NOT_FOUND")
    safe_create_label(toAddr(BASE + 0x7fa8), "STR_BOOT_FILE_WRITE_ERROR")
    safe_create_label(toAddr(BASE + 0x80ca), "STR_BOOT_FLASH_ERASE_ERROR")
    safe_create_label(toAddr(BASE + 0x80e7), "STR_BOOT_FLASH_WRITE_ERROR")
    safe_create_label(toAddr(BASE + 0x8104), "STR_BOOT_FLASH_READ_ERROR")
    safe_create_label(toAddr(BASE + 0x94a0), "STR_FLASH_PROGRESS_ERASE")
    safe_create_label(toAddr(BASE + 0x94b8), "STR_FLASH_PROGRESS_WRITE")


def read_u32(memory, addr):
    data = zeros(4, 'b')
    read = memory.getBytes(addr, data)
    if read != 4:
        return -1
    return ((data[0] & 0xff) << 24) | ((data[1] & 0xff) << 16) | ((data[2] & 0xff) << 8) | (data[3] & 0xff)


def scan_for_refs(memory, start_value, end_value, exact_only):
    hits = []
    for block in memory.getBlocks():
        if not block.isInitialized():
            continue
        cur = block.getStart()
        end = block.getEnd().subtract(3)
        while cur.compareTo(end) <= 0 and not monitor.isCancelled():
            value = read_u32(memory, cur)
            if exact_only:
                match = value == start_value
            else:
                match = value >= start_value and value < end_value
            if match:
                hits.append(cur)
            cur = cur.add(2)
    return hits


def score_candidate(pdis, start, site):
    try:
        cur = start
        count = 0
        covers_site = False
        limit = site.add(8)
        while count < 24 and cur.compareTo(limit) <= 0:
            ins = pdis.disassemble(cur)
            if ins is None:
                return -1
            count += 1
            if ins.getMinAddress().compareTo(site) <= 0 and ins.getMaxAddress().compareTo(site) >= 0:
                covers_site = True
            nxt = ins.getMaxAddress().next()
            if nxt is None or nxt.compareTo(cur) <= 0:
                return -1
            cur = nxt
        if not covers_site:
            return -1
        return (count * 100) + int(site.getOffset() - start.getOffset())
    except:
        return -1


def choose_seed_start(site):
    pdis = PseudoDisassembler(currentProgram)
    best = None
    best_score = -1
    for delta in range(0x40, -2, -2):
        cand = site.subtract(delta)
        score = score_candidate(pdis, cand, site)
        if score > best_score:
            best = cand
            best_score = score
    return best


def disassemble_from(addr):
    if getInstructionAt(addr) is not None:
        return
    cmd = DisassembleCommand(addr, None, True)
    cmd.applyTo(currentProgram, monitor)


def main():
    out_path = None
    args = getScriptArgs()
    if len(args) > 0:
        out_path = args[0]

    memory = currentProgram.getMemory()
    listing = currentProgram.getListing()
    create_anchor_labels()

    raw_hits = {}
    for name, start_value, end_value, exact_only in RANGES:
        raw_hits[name] = scan_for_refs(memory, start_value, end_value, exact_only)

    seed_sites = []
    for addr in raw_hits["progress_block"]:
        seed_sites.append(addr)

    boot_bins = {}
    for addr in raw_hits["boot_update_block"]:
        bin_id = (addr.getOffset() - BASE) // 0x100
        boot_bins[bin_id] = boot_bins.get(bin_id, 0) + 1

    for addr in raw_hits["boot_update_block"]:
        bin_id = (addr.getOffset() - BASE) // 0x100
        if boot_bins.get(bin_id, 0) >= 2:
            seed_sites.append(addr)

    report_lines = []
    report_lines.append("program\t{}".format(currentProgram.getName()))
    report_lines.append("image_base\t0x09000000")
    report_lines.append("boot_update_hits\t{}".format(len(raw_hits["boot_update_block"])))
    report_lines.append("progress_hits\t{}".format(len(raw_hits["progress_block"])))

    for name in raw_hits:
        idx = 0
        for addr in raw_hits[name]:
            safe_create_label(addr, "ref_{}_{}".format(name, idx))
            idx += 1

    unique_starts = set()
    for site in seed_sites:
        start = choose_seed_start(site)
        if start is None:
            start = site
        if start.getOffset() in unique_starts:
            continue
        unique_starts.add(start.getOffset())
        disassemble_from(start)
        func = listing.getFunctionContaining(start)
        if func is None:
            func = createFunction(start, "UPD_{:08x}".format(int(start.getOffset())))
        if func is not None:
            report_lines.append("seed_function\t{}\t{}".format(func.getName(), func.getEntryPoint()))
        else:
            report_lines.append("seed_failed\t{}".format(start))

    if out_path is not None:
        out_file = open(out_path, "w")
        for line in report_lines:
            out_file.write(line + "\n")
        out_file.close()


main()

File: analysis/ghidra_scripts/SeedUpdatePath.java
//@category Analysis

import java.io.File;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import ghidra.app.cmd.disassemble.DisassembleCommand;
import ghidra.app.script.GhidraScript;
import ghidra.app.util.PseudoDisassembler;
import ghidra.app.util.PseudoInstruction;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.mem.Memory;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.program.model.symbol.SourceType;

public class SeedUpdatePath extends GhidraScript {

	private static class RefRange {
		String name;
		long start;
		long end;
		boolean exactOnly;

		RefRange(String name, long start, long end, boolean exactOnly) {
			this.name = name;
			this.start = start;
			this.end = end;
			this.exactOnly = exactOnly;
		}
	}

	private static final long BASE = 0x09000000L;

	private static final RefRange[] RANGES = new RefRange[] {
		new RefRange("boot_update_block", BASE + 0x7be0, BASE + 0x8130, false),
		new RefRange("progress_block", BASE + 0x94a0, BASE + 0x94d0, false),
		new RefRange("progress_erase_exact", BASE + 0x94a0, BASE + 0x94a1, true),
		new RefRange("progress_write_exact", BASE + 0x94b8, BASE + 0x94b9, true),
	};

	@Override
	public void run() throws Exception {
		String outPath = getScriptArgs().length > 0 ? getScriptArgs()[0] : null;
		Listing listing = currentProgram.getListing();
		Memory memory = currentProgram.getMemory();

		createAnchorLabels();

		Map<String, List<Address>> rawHits = new LinkedHashMap<>();
		for (RefRange range : RANGES) {
			rawHits.put(range.name, scanForReferences(memory, range));
		}

		List<Address> seedSites = new ArrayList<>();
		seedSites.addAll(rawHits.get("progress_block"));

		Map<Long, Integer> bootBins = new LinkedHashMap<>();
		for (Address hit : rawHits.get("boot_update_block")) {
			long bin = ((hit.getOffset() - BASE) / 0x100L);
			Integer count = bootBins.get(bin);
			bootBins.put(bin, count == null ? 1 : count + 1);
		}

		for (Address hit : rawHits.get("boot_update_block")) {
			long bin = ((hit.getOffset() - BASE) / 0x100L);
			if (bootBins.get(bin) != null && bootBins.get(bin) >= 2) {
				seedSites.add(hit);
			}
		}

		Set<Long> uniqueSeeds = new LinkedHashSet<>();
		List<String> reportLines = new ArrayList<>();
		reportLines.add("program\t" + currentProgram.getName());
		reportLines.add("image_base\t0x09000000");
		reportLines.add("boot_update_hits\t" + rawHits.get("boot_update_block").size());
		reportLines.add("progress_hits\t" + rawHits.get("progress_block").size());

		for (Map.Entry<String, List<Address>> entry : rawHits.entrySet()) {
			int index = 0;
			for (Address hit : entry.getValue()) {
				String label = "ref_" + entry.getKey() + "_" + index;
				safeCreateLabel(hit, label);
				index++;
			}
		}

		for (Address site : seedSites) {
			Address start = chooseSeedStart(site);
			if (start == null) {
				start = site;
			}
			if (!uniqueSeeds.add(start.getOffset())) {
				continue;
			}
			disassembleFrom(start);
			Function function = listing.getFunctionContaining(start);
			if (function == null) {
				String name = String.format("UPD_%08x", start.getOffset());
				function = createFunction(start, name);
			}
			if (function != null) {
				reportLines.add(String.format("seed_function\t%s\t%s", function.getName(),
					function.getEntryPoint()));
			}
			else {
				reportLines.add(String.format("seed_failed\t%s", start));
			}
		}

		if (outPath != null) {
			writeLines(outPath, reportLines);
		}
	}

	private void createAnchorLabels() throws Exception {
		safeCreateLabel(toAddr(BASE + 0x7c9c), "STR_BOOT_OS_UPDATE_TITLE");
		safeCreateLabel(toAddr(BASE + 0x7cf8), "STR_BOOT_LOADING_OS_FILE");
		safeCreateLabel(toAddr(BASE + 0x7d14), "STR_BOOT_OS_FILE_NOT_FOUND");
		safeCreateLabel(toAddr(BASE + 0x7fa8), "STR_BOOT_FILE_WRITE_ERROR");
		safeCreateLabel(toAddr(BASE + 0x80ca), "STR_BOOT_FLASH_ERASE_ERROR");
		safeCreateLabel(toAddr(BASE + 0x80e7), "STR_BOOT_FLASH_WRITE_ERROR");
		safeCreateLabel(toAddr(BASE + 0x8104), "STR_BOOT_FLASH_READ_ERROR");
		safeCreateLabel(toAddr(BASE + 0x94a0), "STR_FLASH_PROGRESS_ERASE");
		safeCreateLabel(toAddr(BASE + 0x94b8), "STR_FLASH_PROGRESS_WRITE");
	}

	private void safeCreateLabel(Address address, String name) {
		try {
			if (getSymbolAt(address) == null) {
				createLabel(address, name, true, SourceType.ANALYSIS);
			}
		}
		catch (Exception ignored) {
			// Ghidra will throw on duplicate/invalid labels. Ignore and continue.
		}
	}

	private List<Address> scanForReferences(Memory memory, RefRange range) throws Exception {
		List<Address> hits = new ArrayList<>();
		for (MemoryBlock block : memory.getBlocks()) {
			if (!block.isInitialized()) {
				continue;
			}
			Address start = block.getStart();
			long endOffset = block.getEnd().getOffset() - 3;
			for (long off = start.getOffset(); off <= endOffset && !monitor.isCancelled(); off += 2) {
				Address addr = toAddr(off);
				long value = readUnsignedInt(memory, addr);
				boolean match = range.exactOnly ? value == range.start
						: (value >= range.start && value < range.end);
				if (match) {
					hits.add(addr);
				}
			}
		}
		Collections.sort(hits, Comparator.comparingLong(Address::getOffset));
		return hits;
	}

	private long readUnsignedInt(Memory memory, Address address) throws Exception {
		byte[] bytes = new byte[4];
		int read = memory.getBytes(address, bytes);
		if (read != 4) {
			return -1;
		}
		return ((bytes[0] & 0xffL) << 24) | ((bytes[1] & 0xffL) << 16) | ((bytes[2] & 0xffL) << 8) |
			(bytes[3] & 0xffL);
	}

	private Address chooseSeedStart(Address site) throws Exception {
		PseudoDisassembler pdis = new PseudoDisassembler(currentProgram);
		Address best = null;
		int bestScore = -1;
		for (int delta = 0x40; delta >= 0; delta -= 2) {
			Address cand = site.subtract(delta);
			int score = scoreCandidate(pdis, cand, site);
			if (score > bestScore) {
				bestScore = score;
				best = cand;
			}
		}
		return best;
	}

	private int scoreCandidate(PseudoDisassembler pdis, Address start, Address site) {
		try {
			Address cur = start;
			int count = 0;
			boolean coversSite = false;
			while (count < 24 && cur.compareTo(site.add(8)) <= 0) {
				PseudoInstruction ins = pdis.disassemble(cur);
				if (ins == null) {
					return -1;
				}
				count++;
				Address next = ins.getMaxAddress().next();
				if (next == null || next.compareTo(cur) <= 0) {
					return -1;
				}
				if (ins.getMinAddress().compareTo(site) <= 0 &&
					ins.getMaxAddress().compareTo(site) >= 0) {
					coversSite = true;
				}
				cur = next;
			}
			if (!coversSite) {
				return -1;
			}
			return (count * 100) + (int) (site.getOffset() - start.getOffset());
		}
		catch (Exception e) {
			return -1;
		}
	}

	private void disassembleFrom(Address start) throws Exception {
		Instruction existing = getInstructionAt(start);
		if (existing != null) {
			return;
		}
		DisassembleCommand cmd = new DisassembleCommand(start, null, true);
		cmd.applyTo(currentProgram, monitor);
	}

	private void writeLines(String outPath, List<String> lines) throws Exception {
		File outFile = new File(outPath);
		outFile.getParentFile().mkdirs();
		PrintWriter writer = new PrintWriter(outFile);
		try {
			for (String line : lines) {
				writer.println(line);
			}
		}
		finally {
			writer.close();
		}
	}
}
